### PR TITLE
[DataPipe] Align shuffling behavior for IterDataPipe and MapDataPipe

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82975 [DataLoader] Add shuffling logic for MapDataPipe
* **#82974 [DataPipe] Align shuffling behavior for IterDataPipe and MapDataPipe**

This PR align the shuffling behavior of two types of `DataPipe`:
- Shuffling will be lazily executed only after `__iter__` or `__getitem__` is invoked.
- [Discussion] `seed` will be fixed unless `set_seed` is invoked.
  - BC breaking for `IterDataPipe` as the original behavior would a seed will be generated whenever `iter` is invoked, but now each `iterator` would share the same seed until `set_seed` is invoked.
- Tests exist as `test_shuffler_iterdatapipe` and `test_shuffler_mapdatapipe`

### CODE REVIEW

**Review by NivekT — COMMENTED** [review]
Overall, I agree that we should align the shuffle behaviors between two styles of DataPipe.

Letting DataLoader control this behavior makes sense. At the expense of UX of people who don't use DataPipes with DataLoader for whatever reason.

1. I do wonder if we should allow users to toggle this behavior (same or different seed with new iterator) on and off.
2. We need to update all the docstring and documentation to make sure that the behavior is clear to users.

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 1]
```diff
@@ -116,14 +116,16 @@ def set_shuffle(self, shuffle=True):
 
     def set_seed(self, seed: int):
         self._seed = seed
+        return self
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py:119]: I think this is general syntax sugar for users:
```
dp = dp.set_seed(1)
dp = dp.set_shuffle(True)
```
It prevents users accidentally converting `dp` to `None`.

File: test/test_datapipe.py [hunk 2]
```diff
@@ -1554,7 +1554,7 @@ def test_stream_reader_iterdatapipe(self):
         with self.assertRaises(TypeError):
             len(dp1)
 
-    def test_shuffle_iterdatapipe(self):
+    def test_shuffler_iterdatapipe(self):
```
[the remark this answers is not in the document]
Comment (ejguan) [reply test/test_datapipe.py:1557]: Let me re-construct the whole tests for shuffle.

File: torch/utils/data/datapipes/map/combinatorics.py [hunk 3]
```diff
@@ -45,15 +46,37 @@ def __init__(self,
...
+        self._reset = True
+        return self
 
     def __getitem__(self, index) -> T_co:
         try:
             old_numeric_index = self.index_map[index]
         except KeyError:
             raise IndexError(f"Index {index} is out of range for {self}.")
-        new_index = self.indices[old_numeric_index]
+
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/utils/data/datapipes/map/combinatorics.py:69]: This will be introduced by the next PR.

**Comment by ejguan** [comment]
Thinking for a while on `MapDataPipe.shuffle` today, I find it becomes anti-intuitive to me that `Shuffler` is a still a `MapDataPipe`. We normally refer `MapDataPipe` as we provide an index to this instance, it gives a corresponding object. However, shuffling would make it so weird that a non-corresponding object is returned given an index. I personally think it makes more sense that `MapDataPipe` should be converted to an `IterDataPipe` after shuffling.

**Comment by ejguan** [comment]
Closing this PR as we all agreed on moving `MapDataPipe.shuffle` to an `IterDataPipe` during an offline discussion.

**Pull request closed without merging**